explainability-service/src/main/java/org/kie/trustyai/service/endpoints/metrics/drift/ApproxKSTestEndpoint.java

```diff
@@ -0,0 +1,106 @@
+    }
+
+    // a specific definition for this value of this metric in this specific context
+    @Override
+    public String getSpecificDefinition(MetricValueCarrier metricValues, @ValidDriftMetricRequest ApproxKSTestMetricRequest request) {
+        StringBuilder out = new StringBuilder(getGeneralDefinition());
+        out.append(System.getProperty("line.separator"));
+
+        int maxColLen = metricValues.getNamedValues().keySet().stream().mapToInt(s -> s.length()).max().orElse(0);
+        String fmt = String.format("%%%ds", maxColLen);
+
+        for (Map.Entry<String, Double> entry : metricValues.getNamedValues().entrySet()) {
+            boolean reject = entry.getValue() <= request.getThresholdDelta();
+            out.append(String.format("  - Column %s has p=%f probability of coming from the training distribution.",
+                    String.format(fmt, entry.getKey()),
+                    entry.getValue()));
+            if (reject) {
+                out.append(String.format(" p <= %f -> [SIGNIFICANT DRIFT]", request.getThresholdDelta()));
+            } else {
+                out.append(String.format(" p >  %f", request.getThresholdDelta()));
+            }
+            out.append(System.getProperty("line.separator"));
+        }
+        return out.toString();
+    }
+
+    @Override
+    @CacheResult(cacheName = "metrics-calculator-approxkstest", keyGenerator = MetricCalculationCacheKeyGen.class)
+    public MetricValueCarrier calculate(Dataframe dataframe, @ValidReconciledMetricRequest BaseMetricRequest request) {
+        ApproxKSTestMetricRequest aksRequest = (ApproxKSTestMetricRequest) request;
+        ApproxKSFitting aksFitting;
+        if (aksRequest.getSketchFitting() == null) {
+            LOG.debug("Fitting a approxKSTest drift request for model=" + request.getModelId());
+            // get the data that matches the provided reference tag: calibration data
+            Dataframe fitting = super.dataSource.get()
+                    .getDataframe(request.getModelId())
+                    .filterRowsByTagEquals(aksRequest.getReferenceTag());
+            aksFitting = ApproxKSTest.precompute(fitting, aksRequest.getEpsilon() );
+            aksRequest.setSketchFitting(aksFitting.getfitSketches());
+        } else {
+            LOG.debug("Using previously found fitting data sketches in request for model=" + request.getModelId());
+            aksFitting = new ApproxKSFitting(aksRequest.getSketchFitting());
+        }
+        ApproxKSTest approxKS = new ApproxKSTest(aksFitting, aksRequest.getEpsilon());
+        LOG.debug("Cache miss. Calculating metric for " + aksRequest.getModelId());
+        // get data that does _not_ have the provided reference tag: test data
+        Dataframe filtered = dataframe.filterRowsByTagNotEquals(aksRequest.getReferenceTag());
+        Map<String, HypothesisTestResult> result = approxKS.calculate(filtered, aksRequest.getThresholdDelta());
```

Comment:
We need to safeguard against the case where`dataframe.getRowDimension < 2`. Right now, we'd get an IllegalArgumentException, but we need a non-erroring path. For MeanshiftEndpoint, it returns a p=1 for all columns, the reasoning being that if there are <2 datapoints the statistical test is as insignificant as possible.